backend/repository/validator.go

```diff
@@ -0,0 +1,378 @@
+	return visitor.Visit()
+}
+
+// ResolveNode resolves a node in thrift AST given a string path
+// example: [SimpleService, compare]
+func ResolveNode(nodePaths []string, root *compile.Module) (interface{}, error) {
+	var node interface{}
+	node = root
+	for _, n := range nodePaths {
+		switch node.(type) {
+		case *compile.Module:
+			m := node.(*compile.Module)
+			n, ok := m.Services[n]
+			if !ok {
+				return nil, fmt.Errorf("unable to resolve path %s", nodePaths)
+			}
+			node = n
+			break
+		case *compile.ServiceSpec:
+			s := node.(*compile.ServiceSpec)
+			n, ok := s.Functions[n]
+			if !ok {
+				return nil, fmt.Errorf("unable to resolve path %s", nodePaths)
+			}
+			node = n
+			break
+		default:
+			return nil, fmt.Errorf("unable to resolve path %s", nodePaths)
+		}
+	}
+	return node, nil
+}
+
+// Visitor provides interface to visits AST
+type Visitor interface {
+	Visit() error
+
+	VisitService(*compile.ServiceSpec, *compile.ServiceSpec) error
+	VisitFunction(*compile.FunctionSpec, *compile.FunctionSpec) error
+	VisitArgsSpec(compile.ArgsSpec, compile.ArgsSpec) error
+	VisitResultSpec(*compile.ResultSpec, *compile.ResultSpec) error
+	VisitFieldSpec(*compile.FieldSpec, *compile.FieldSpec) error
+	VisitFieldGroup(compile.FieldGroup, compile.FieldGroup) error
+	VisitTypeSpec(compile.TypeSpec, compile.TypeSpec) error
+}
+
+// NodePair include client and endpoint nodes used to compare
+type NodePair struct {
+	ClientNode   interface{}
+	EndpointNode interface{}
+}
+
+// NodeQ is a FIFO queue for NodePair
+type NodeQ struct {
+	q []NodePair
+}
+
+// NewNodeQ returns a NodeQ
+func NewNodeQ() *NodeQ {
+	return &NodeQ{
+		q: make([]NodePair, 0),
+	}
+}
+
+// EnQ enques a pair of client/endpoint to compare
+func (nq *NodeQ) EnQ(client, endpoint interface{}) {
+	node := NodePair{
+		ClientNode:   client,
+		EndpointNode: endpoint,
+	}
+	nq.q = append(nq.q, node)
+}
+
+// DeQ dequeues a pair of client/endpoint to compare from head of the queue
+func (nq *NodeQ) DeQ() (client, endpoint interface{}, err error) {
+	if nq.Empty() {
+		err = errors.New("empty queue")
+		return
+	}
+	node := nq.q[0]
+	client = node.ClientNode
+	endpoint = node.EndpointNode
+	nq.q = nq.q[1:]
+	return
+}
+
+// Empty returns true if queue is empty
+func (nq NodeQ) Empty() bool {
+	return len(nq.q) == 0
+}
+
+// SyntaxVisitor implements a syntax compare visitor interface
+type SyntaxVisitor struct {
+	nodeQ *NodeQ
+}
+
+// NewSyntaxVisitor returns a SyntaxVisitor
+func NewSyntaxVisitor(clientNode, endpointNode interface{}) *SyntaxVisitor {
+	nodeQ := NewNodeQ()
+	nodeQ.EnQ(clientNode, endpointNode)
+	return &SyntaxVisitor{
+		nodeQ: nodeQ,
+	}
+}
+
+// Visit BFS visits all node pairs in node queue
+func (sv *SyntaxVisitor) Visit() error {
+	var (
+		client   interface{}
+		endpoint interface{}
+		err      error
+	)
+	for {
+		client, endpoint, err = sv.nodeQ.DeQ()
+		if err != nil {
+			return nil
+		}
+		if reflect.TypeOf(client) != reflect.TypeOf(endpoint) {
```

Comment:
I don't think this hold true when we implement rewrite rules ( with transform middleware ).